File: 23-merge-k-sorted-lists/23-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        
        heap = []
        
        for idx, node in enumerate(lists):
            if node:
                heap.append((node.val, idx, node))
            
        heapq.heapify(heap)
        
        head = ListNode()
        curr = head
        
        while heap:
            small = heapq.heappop(heap)
            curr.next = small[2]
            curr = curr.next
            
            if curr.next:
                demo = curr.next
                heapq.heappush(heap,(demo.val, small[1], demo))
      
        return head.next

File: 23-merge-k-sorted-lists/23-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode* ans = new ListNode(0);
        ListNode* cur = ans;
        priority_queue<pair<int, ListNode*>> pq;
        
        for (auto L : lists)
            if (L) pq.emplace(-L->val, L);
        
        while (!pq.empty()) {
            auto [val, node] = pq.top();
            pq.pop();
            ans->next = node;
            ans = ans->next;
            if (node->next) {
                node = node->next;
                pq.emplace(-node->val, node);
            }
        }
        return cur->next;
    }
};